when cluster is not available should build not fail when deploy requested ?

**Describe the bug**
have cluster configured in kube but the cluster being down/not reachable:
run `mvn clean package -Dquarkus.kubernetes.deploy=true`

and you get output like this:

```shell
[WARNING] Error reading service account token from: [/var/run/secrets/kubernetes.io/serviceaccount/token]. Ignoring.
[WARNING] Error reading service account token from: [/var/run/secrets/kubernetes.io/serviceaccount/token]. Ignoring.
[ERROR] [io.quarkus.kubernetes.deployment.KubernetesDeploy] Although a Kubernetes deployment was requested, it will however not take place because there was an error during communication with the API Server: An error has occurred.
[WARNING] Error reading service account token from: [/var/run/secrets/kubernetes.io/serviceaccount/token]. Ignoring.
[WARNING] Error reading service account token from: [/var/run/secrets/kubernetes.io/serviceaccount/token]. Ignoring.
[INFO] [org.jboss.threads] JBoss Threads version 3.1.1.Final
[WARNING] Error reading service account token from: [/var/run/secrets/kubernetes.io/serviceaccount/token]. Ignoring.
[WARNING] Error reading service account token from: [/var/run/secrets/kubernetes.io/serviceaccount/token]. Ignoring.
[INFO] [io.quarkus.deployment.pkg.steps.JarResultBuildStep] Building thin jar: /Users/max/Dropbox/code/openshift-deployment/target/simple-1.0.0-SNAPSHOT-runner.jar
[INFO] Initializing dekorate session.
[INFO] Default s2i build generator....
[INFO] Generating manifests.
[INFO] Processing kubernetes configuration.
[INFO] Processing openshift configuration.
[INFO] Processing s2i configuration.
[INFO] Closing dekorate session.
[INFO] [io.quarkus.deployment.QuarkusAugmentor] Quarkus augmentation completed in 8761ms
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
```


**Expected behavior**
I would have expected a BUILD FAILURE here.

**Actual behavior**
build marked as success despite there are errors.

--- a/extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesDeploy.java
+++ b/extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesDeploy.java
@@ -3,7 +3,7 @@ package io.quarkus.kubernetes.deployment;
 
 import static io.quarkus.kubernetes.deployment.Constants.DEPLOY;
 
-import java.util.function.BooleanSupplier;
+import java.util.Optional;
 
 import javax.net.ssl.SSLHandshakeException;
 
@@ -15,42 +15,108 @@ import io.fabric8.kubernetes.api.model.RootPaths;
 import io.fabric8.kubernetes.client.KubernetesClient;
 import io.quarkus.kubernetes.client.runtime.KubernetesClientUtils;
 
-public class KubernetesDeploy implements BooleanSupplier {
+public class KubernetesDeploy {
+
+    public static KubernetesDeploy INSTANCE = new KubernetesDeploy();
 
     private static final Logger log = Logger.getLogger(KubernetesDeploy.class);
     private static boolean serverFound = false;
-    private static boolean alreadyWarned = false;
 
-    @Override
-    public boolean getAsBoolean() {
+    private KubernetesDeploy() {
+    }
+
+    /**
+     * @return {@code true} iff @{code quarkus.kubernetes.deploy=true} AND the target Kubernetes API server is reachable
+     *
+     *         It follows that this can only return {@code false} if the if @{code quarkus.kubernetes.deploy=false}
+     *
+     * @throws RuntimeException if communication to the Kubernetes API server errored
+     */
+    public boolean check() {
+        Result result = doCheck();
+
+        if (result.getException().isPresent()) {
+            throw result.getException().get();
+        }
+
+        return result.isAllowed();
+    }
+
+    /**
+     * @return {@code true} iff @{code quarkus.kubernetes.deploy=true} AND the target Kubernetes API server is reachable
+     *
+     *         Never throws an exception even in the face of a communication error with the API server, just returns
+     *         {@code false}
+     *         in that case
+     */
+    public boolean checkSilently() {
+        return doCheck().isAllowed();
+    }
+
+    private Result doCheck() {
         Config config = ConfigProvider.getConfig();
         if (!config.getOptionalValue(DEPLOY, Boolean.class).orElse(false)) {
-            return false;
+            return Result.notConfigured();
         }
 
         // No need to perform the check multiple times.
         if (serverFound) {
-            return true;
+            return Result.enabled();
         }
-        try (final KubernetesClient client = KubernetesClientUtils.createClient()) {
+
+        KubernetesClient client = KubernetesClientUtils.createClient();
+        String masterURL = client.getConfiguration().getMasterUrl();
+        try {
+            masterURL = client.getConfiguration().getMasterUrl();
             //Let's check id we can connect.
             RootPaths paths = client.rootPaths();
-            log.info("Found kubernetes server.");
+            log.info("Kubernetes API Server at '" + masterURL + "' successfully contacted.");
             serverFound = true;
-            return true;
+            client.close();
+            return Result.enabled();
         } catch (Exception e) {
-            if (!alreadyWarned) {
-                if (e.getCause() instanceof SSLHandshakeException) {
-                    String message = "Although a Kubernetes deployment was requested, it will however not take place because the API Server certificates are not trusted. The certificates can be configured using the relevant configuration propertiers under the 'quarkus.kubernetes-client' config root, or \"quarkus.kubernetes-client.trust-certs=true\" can be set to explicitly trust the certificates (not recommended)";
-                    log.warn(message);
-                } else {
-                    log.error(
-                            "Although a Kubernetes deployment was requested, it will however not take place because there was an error during communication with the API Server: "
-                                    + e.getMessage());
-                }
-                alreadyWarned = true;
+            if (e.getCause() instanceof SSLHandshakeException) {
+                return Result.exceptional(new RuntimeException(
+                        "Although a Kubernetes deployment was requested, it however cannot take place because the API Server (at '"
+                                + masterURL
+                                + "') certificates are not trusted. The certificates can be configured using the relevant configuration propertiers under the 'quarkus.kubernetes-client' config root, or \"quarkus.kubernetes-client.trust-certs=true\" can be set to explicitly trust the certificates (not recommended)",
+                        e));
+            } else {
+                return Result.exceptional(new RuntimeException(
+                        "Although a Kubernetes deployment was requested, it however cannot take place because there was an error during communication with the API Server at '"
+                                + masterURL + "'",
+                        e));
             }
-            return false;
+        }
+    }
+
+    private static class Result {
+        private final boolean allowed;
+        private final Optional<RuntimeException> exception;
+
+        private Result(boolean allowed, Optional<RuntimeException> exception) {
+            this.allowed = allowed;
+            this.exception = exception;
+        }
+
+        static Result notConfigured() {
+            return new Result(false, Optional.empty());
+        }
+
+        static Result enabled() {
+            return new Result(true, Optional.empty());
+        }
+
+        static Result exceptional(RuntimeException e) {
+            return new Result(false, Optional.of(e));
+        }
+
+        public boolean isAllowed() {
+            return allowed;
+        }
+
+        public Optional<RuntimeException> getException() {
+            return exception;
         }
     }
 }

--- a/extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesDeployer.java
+++ b/extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesDeployer.java
@@ -47,7 +47,7 @@ public class KubernetesDeployer {
             .values().stream()
             .map(s -> "\"" + s + "\"").collect(Collectors.joining(", "));
 
-    @BuildStep(onlyIf = { IsNormal.class, KubernetesDeploy.class })
+    @BuildStep(onlyIf = IsNormal.class)
     public void deploy(KubernetesClientBuildItem kubernetesClient,
             ContainerImageInfoBuildItem containerImageInfo,
             EnabledKubernetesDeploymentTargetsBuildItem targets,
@@ -56,6 +56,10 @@ public class KubernetesDeployer {
             BuildProducer<DeploymentResultBuildItem> deploymentResult,
             // needed to ensure that this step runs after the container image has been built
             @SuppressWarnings("unused") List<ArtifactResultBuildItem> artifactResults) {
+
+        if (!KubernetesDeploy.INSTANCE.check()) {
+            return;
+        }
 
         Optional<String> activeContainerImageCapability = ContainerImageCapabilitiesUtil
                 .getActiveContainerImageCapability(capabilities);

--- a/extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesDeployerPrerequisite.java
+++ b/extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesDeployerPrerequisite.java
@@ -1,21 +1,25 @@
 package io.quarkus.kubernetes.deployment;
 
-import org.jboss.logging.Logger;
-
 import io.quarkus.container.spi.ContainerImageBuildRequestBuildItem;
 import io.quarkus.container.spi.ContainerImageInfoBuildItem;
 import io.quarkus.container.spi.ContainerImagePushRequestBuildItem;
+import io.quarkus.deployment.IsNormal;
 import io.quarkus.deployment.annotations.BuildProducer;
 import io.quarkus.deployment.annotations.BuildStep;
 
 public class KubernetesDeployerPrerequisite {
 
-    private static final Logger log = Logger.getLogger(KubernetesDeploy.class);
-
-    @BuildStep(onlyIf = KubernetesDeploy.class)
+    @BuildStep(onlyIf = IsNormal.class)
     public void prepare(ContainerImageInfoBuildItem containerImage,
             BuildProducer<ContainerImageBuildRequestBuildItem> buildRequestProducer,
             BuildProducer<ContainerImagePushRequestBuildItem> pushRequestProducer) {
+
+        // we don't want to throw an exception at this step and fail the build because it could prevent
+        // the Kubernetes resources from being generated
+        if (!KubernetesDeploy.INSTANCE.checkSilently()) {
+            return;
+        }
+
         //Let's communicate to the container-image plugin that we need an image build and an image push.
         buildRequestProducer.produce(new ContainerImageBuildRequestBuildItem());
         if (containerImage.getRegistry().isPresent()) {

--- a/integration-tests/kubernetes/maven-invoker-way/src/it/kubernetes-docker-build-and-deploy/invoker.properties
+++ b/integration-tests/kubernetes/maven-invoker-way/src/it/kubernetes-docker-build-and-deploy/invoker.properties
@@ -1,2 +1,4 @@
 # invoker.goals=clean package -Dquarkus.container.build=true -Dquarkus.package.type=native 
 invoker.goals=clean package -Dquarkus.kubernetes.deploy=true
+# expect a failure since there is no Kubernetes cluster to deploy to
+invoker.buildResult = failure

--- a/integration-tests/kubernetes/maven-invoker-way/src/it/kubernetes-jib-build-and-deploy/invoker.properties
+++ b/integration-tests/kubernetes/maven-invoker-way/src/it/kubernetes-jib-build-and-deploy/invoker.properties
@@ -1,2 +1,4 @@
 # invoker.goals=clean package -Dquarkus.container.build=true -Dquarkus.package.type=native 
 invoker.goals=clean package -Dquarkus.kubernetes.deploy=true
+# expect a failure since there is no Kubernetes cluster to deploy to
+invoker.buildResult = failure

--- a/integration-tests/kubernetes/maven-invoker-way/src/it/openshift-s2i-build-and-deploy/invoker.properties
+++ b/integration-tests/kubernetes/maven-invoker-way/src/it/openshift-s2i-build-and-deploy/invoker.properties
@@ -1,2 +1,4 @@
 # invoker.goals=clean package -Dquarkus.container.build=true -Dquarkus.package.type=native 
 invoker.goals=clean package -Dquarkus.kubernetes.deploy=true
+# expect a failure since there is no Kubernetes cluster to deploy to
+invoker.buildResult = failure

--- a/integration-tests/kubernetes/quarkus-standard-way/src/test/java/io/quarkus/it/kubernetes/DeserializationUtil.java
+++ b/integration-tests/kubernetes/quarkus-standard-way/src/test/java/io/quarkus/it/kubernetes/DeserializationUtil.java
@@ -16,7 +16,7 @@ import io.fabric8.kubernetes.api.model.HasMetadata;
 final class DeserializationUtil {
 
     private static final String DOCUMENT_DELIMITER = "---";
-    private static final ObjectMapper MAPPER = new ObjectMapper(new YAMLFactory());
+    static final ObjectMapper MAPPER = new ObjectMapper(new YAMLFactory());
 
     private DeserializationUtil() {
     }
@@ -39,7 +39,7 @@ final class DeserializationUtil {
         return items;
     }
 
-    private static String[] splitDocument(String[] lines) {
+    static String[] splitDocument(String[] lines) {
         List<String> documents = new ArrayList<>();
         int nLine = 0;
         StringBuilder builder = new StringBuilder();

--- a/integration-tests/kubernetes/quarkus-standard-way/src/test/java/io/quarkus/it/kubernetes/KnativeContainerImageTest.java
+++ b/integration-tests/kubernetes/quarkus-standard-way/src/test/java/io/quarkus/it/kubernetes/KnativeContainerImageTest.java
@@ -2,10 +2,12 @@
 package io.quarkus.it.kubernetes;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.fail;
 
 import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
 import java.nio.file.Path;
-import java.util.List;
 
 import org.jboss.shrinkwrap.api.ShrinkWrap;
 import org.jboss.shrinkwrap.api.spec.JavaArchive;
@@ -13,7 +15,6 @@ import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.RegisterExtension;
 
 import io.fabric8.knative.serving.v1.Service;
-import io.fabric8.kubernetes.api.model.HasMetadata;
 import io.quarkus.test.ProdBuildResults;
 import io.quarkus.test.ProdModeTestResults;
 import io.quarkus.test.QuarkusProdModeTest;
@@ -36,22 +37,29 @@ public class KnativeContainerImageTest {
                 .isDirectoryContaining(p -> p.getFileName().endsWith("knative.json"))
                 .isDirectoryContaining(p -> p.getFileName().endsWith("knative.yml"));
 
-        List<HasMetadata> kubernetesList = DeserializationUtil.deserializeAsList(kubernetesDir.resolve("knative.yml"));
-
-        assertThat(kubernetesList).filteredOn(i -> "Service".equals(i.getKind())).hasOnlyOneElementSatisfying(i -> {
-            assertThat(i).isInstanceOfSatisfying(Service.class, d -> {
-
-                assertThat(d.getSpec()).satisfies(serviceSpec -> {
-                    assertThat(serviceSpec.getTemplate()).satisfies(revisionTemplate -> {
-                        assertThat(revisionTemplate.getSpec()).satisfies(spec -> {
-                            assertThat(spec.getContainers()).satisfies(containers -> {
-                                assertThat(containers.get(0)).satisfies(c -> assertThat(c.getImage())
-                                        .isEqualTo("quay.io/grp/knative-with-container-image:0.1-SNAPSHOT"));
-                            });
-                        });
+        assertThat(getKNativeService(kubernetesDir).getSpec()).satisfies(serviceSpec -> {
+            assertThat(serviceSpec.getTemplate()).satisfies(revisionTemplate -> {
+                assertThat(revisionTemplate.getSpec()).satisfies(spec -> {
+                    assertThat(spec.getContainers()).satisfies(containers -> {
+                        assertThat(containers.get(0)).satisfies(c -> assertThat(c.getImage())
+                                .isEqualTo("quay.io/grp/knative-with-container-image:0.1-SNAPSHOT"));
                     });
                 });
             });
         });
+    }
+
+    private Service getKNativeService(Path kubernetesDir) throws IOException {
+        String[] yamlFiles = DeserializationUtil.splitDocument(
+                Files.readAllLines(kubernetesDir.resolve("knative.yml"), StandardCharsets.UTF_8).toArray(new String[0]));
+
+        for (String yamlFile : yamlFiles) {
+            if (yamlFile.contains("\"Service\"")) {
+                return DeserializationUtil.MAPPER.readValue(yamlFile, Service.class);
+            }
+        }
+
+        fail("No KNative Service was generated");
+        return null; // keep the compiler happy
     }
 }

